### ISSUE TITLE
[RFC] Tuple unpacking in loops/iterators

### ISSUE THREAD

**Issue by narimiran** [nim-lang/Nim#7486]
## Current situation

To iterate through a container, we now have:
```nim
for val in mySeq:
  ...
```

There is also an option to have an index too with:
```nim
for i, val in mySeq:
  ...
```

If `mySeq` is container of tuples, we currently need to do:
```nim
for val in mySeq:
  let
    x = val[0]
    y = val[1]
    z = val[2]
  ...
```

An example of `zip` usage:
```nim
for val in zip(seqA, seqB):
  let area = val.a * val.b # you must use `a` and `b`, no other letters/words
```

----

## Proposal

My proposition is to have tuple unpacking available in loops, and to use the same syntax which is currently used for tuple unpacking in `let` statements: `let (x, y, z) = myTuple`.

The above example now looks like this:
```nim
for (x, y, z) in mySeq:
  ...
```

or if you want to also use an index:
```nim
for i, (x, y, z) in mySeq:
  ...
```

The example with `zip` would now allow more meaningful names:
```nim
for (length, height) in zip(seqA, seqB):
  let area = length * height
```

---

Parentheses must be used to make a clear, understandable difference between `for i, val in mySeq` and `for (x, y) in mySeq`, where `i` is an index, `val` is a whole tuple, and `x` and `y` are elements of that tuple.

This would allow to have a cleaner, nicer and more readable code.

Thoughts?

**Comment by dom96** [comment]
Sounds good to me.

👍

**Comment by Yardanico** [comment]
@narimiran about 
```nim
for val in mySeq:
  let
    x = val[0]
    y = val[1]
    z = val[2]
  ...
```
It's not the best, but it can be a little bit easier:
```nim
let mySeq = @[(1, 2, 3), (3, 4, 5), (3, 1, 2)]
for val in mySeq:
  let (x, y, z) = val
```

**Comment by krux02** [comment]
well this is called pattern matching. And I think that full pattern matching should be supported in nim

**Comment by narimiran** [comment]
> a full pattern matching library should be the goal

Agreed. Or even better yet - not a library, but a standard Nim feature.

But until that happens, tuple unpacking as proposed would be a nice thing to have.

**Comment by andreaferretti** [comment]
Something similar exists in the form of https://github.com/numforge/loop-fusion

**Comment by narimiran** [comment]
@andreaferretti Correct me if I'm wrong, but to me it looks like Loop Fusion solves the problem of `zip` unpacking, but it cannot be used for unpacking of `seq[tuple]` (cc @mratsim)

**Comment by notTito** [comment]
-1 from me. You can write:
```
for a, b, c in items(mySeq):
   ...
```

**Comment by krux02** [comment]
@notTito the problem is that it doesn't work when you also want the index.

**Comment by Yardanico** [comment]
@notTito but it doesn't work without calling `items` or when you need the index too

**Comment by andreaferretti** [comment]
@narimiran You are right, I mentioned it because I think the same approach could be extended to tuples

**Comment by notTito** [comment]
Ok you guys are are right.

**Comment by narimiran** [comment]
Now I've seen @dom96 had similar idea back in 2015: #2436

This received quite an amount of thumb-ups — can we have this before v1.0?

**Comment by dom96** [comment]
This isn't necessary for v1.0

**Comment by mratsim** [comment]
Note that the difference between iterators that return tuples and seq of tuples must be documented.

In the first case, iterators like pairs don't need parenthesis today:

```Nim
iterator myPair[T](x: seq[T]): (int, T) =

  for i in 0 ..< x.len:
    yield (i, x[i])  # Packed tuple


let pow2 = @[1, 2, 4, 8, 16, 32]

for j, val in pow2.myPair(): # magically unpacked
  echo "2^", j, " = ", val
```
